csharp/src/Ice/LocatorInfo.cs

```diff
@@ -169,148 +166,164 @@ internal void ClearCache(Reference reference)
                     {
                         // Endpoints are returned from the cache but TTL was reached, if backgrounds updates
                         // are configured, we obtain new endpoints but continue using the stale endpoints to
                         // not block the caller.
-                        _ = GetAdapterEndpointsAsync(reference, cancel: default);
+                        _ = ResolveLocationAsync(reference, cancel: default);
                     }
                     else
                     {
-                        endpoints = await GetAdapterEndpointsAsync(reference, cancel).ConfigureAwait(false);
+                        directReference =
+                            await ResolveLocationAsync(reference, cancel).ConfigureAwait(false);
                     }
                 }
             }
 
             if (reference.Communicator.TraceLevels.Location >= 1)
             {
-                if (endpoints.Count > 0)
+                if (directReference != null)
                 {
                     if (cached)
                     {
                         if (reference.IsWellKnown)
                         {
-                            Trace("found endpoints for well-known proxy in locator cache", reference, endpoints);
+                            TraceDirect("found entry for well-known proxy in locator cache",
+                                        reference,
+                                        directReference);
                         }
                         else
                         {
-                            Trace("found endpoints for adapter in locator cache", reference, endpoints);
+                            TraceDirect("found entry for location in locator cache", reference, directReference);
                         }
                     }
                     else
                     {
                         if (reference.IsWellKnown)
                         {
-                            Trace("retrieved endpoints for well-known proxy from locator, adding to locator cache",
-                                  reference,
-                                  endpoints);
+                            TraceDirect("resolved well-known proxy using locator, adding to locator cache",
+                                         reference,
+                                         directReference);
                         }
                         else
                         {
-                            Trace("retrieved endpoints for adapter from locator, adding to locator cache",
-                                  reference,
-                                  endpoints);
+                            TraceDirect("resolved location using locator, adding to locator cache",
+                                        reference,
+                                        directReference);
                         }
                     }
```

Comment:
maybe we can do this tracing a bit more compact

```
 var sb = new System.Text.StringBuilder();
 sb.Append(cached ? "found entry for " : "resolved ");
 sb.Append(reference.IsWellKnown ? "well-known proxy " : "location ");
 sb.Append(cached ? "in locator cache" : "using locator, adding to locator cache");
 TraceDirect(sb.ToString(), reference, directReference);
```

Reply:
Reworked this code.